prow/gangway/gangway.go

```diff
@@ -0,0 +1,759 @@
+	jobs, err := pc.GetPostsubmits(orgRepo, baseSHAGetter)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]interface{}, len(jobs))
+	for i, v := range jobs {
+		ret[i] = v
+	}
+	return ret, nil
+}
+
+// FetchJobStruct looks at the sea of all possible Prow Job definitions and
+// selects The One that matches the details in the given JobDefinition (`jd`).
+func (gw *Gangway) FetchJobStruct(req *CreateJobExecutionRequest) (*JobStruct, error) {
+	// We need to now write a single getProwJobSpec "handler" function
+	// that handles all 3 job execution types. In the pub/sub code we do this
+	// with 3 separate functions with a certain amount of code duplication
+	// across them, but we just do it here in one function for simplicity.
+
+	jobDefinition := req.GetJobDefinition()
+	jobName := jobDefinition.GetName()
+
+	// Only used for postsubmits and presubmits.
+	baseRepo := req.GetBaseRepo()
+	baseRepoCommit := baseRepo.GetBase().GetCommit()
+	baseRepoRef := baseRepo.GetBase().GetRef()
+	baseSHAGetter := func() (string, error) {
+		return baseRepoCommit, nil
+	}
+
+	// Only used for presubmits.
+	var headSHAGetters []func() (string, error)
+	refsToMerge := baseRepo.GetRefsToMerge()
+	for _, refToMerge := range refsToMerge {
+		refToMerge := refToMerge
+		headSHAGetters = append(headSHAGetters, func() (string, error) {
+			return refToMerge.GetCommit(), nil
+		})
+	}
+
+	jobStruct := JobStruct{}
+	jobStruct.JobExecutionType = jobDefinition.GetExecutionType()
+
+	cfg := gw.ConfigAgent.Config()
+	pc := gw.InRepoConfigCacheHandler
+
+	var jobFetchers jobFetchers
+
+	switch jobStruct.JobExecutionType {
+	case JobExecutionType_PERIODIC:
+		// Search for the correct Periodic job from the possible candidates
+		// defined in the central repo.
+		for _, candidateJob := range cfg.AllPeriodics() {
+			candidateJob := candidateJob
+			if candidateJob.Name == jobDefinition.GetName() {
+				(*jobStruct.Job) = &candidateJob
+				break
+			}
+		}
+		if jobStruct.Job == nil {
+			return nil, fmt.Errorf("failed to find associated periodic job %q", jobDefinition.GetName())
+		}
+	case JobExecutionType_POSTSUBMIT:
+		jobFetchers = &postsubmitFetchers{}
+	case JobExecutionType_PRESUBMIT:
+		jobFetchers = &presubmitFetchers{}
+	}
+
+	// Handle presubmits and postsubmits with the same fetching logic.
+	switch jobStruct.JobExecutionType {
+	case JobExecutionType_POSTSUBMIT:
+		fallthrough
+	case JobExecutionType_PRESUBMIT:
+		// FIXME: Need to strip the leading "https://github.com/" for GitHub, and add an integration test.
+		org, repo, err := getOrgRepo(baseRepo.GetBase().GetUrl())
+		if err != nil {
+			return nil, err
+		}
+		orgRepo := strings.Join([]string{org, repo}, "/")
+
+		// Fetching the statically-defined postsubmit jobs requires providing an "orgRepo" filter.
+		var jobs []interface{}
+
+		if jobDefinition.GetDefinitionType() == JobDefinitionType_CENTRAL_REPO {
```

Comment:
@cjwagner I think having this JobDefinitionType_... check is useful here.

Reply:
Why do we need to make this distinction at all? Whether or not jobs come from the static config or inrepo config is already abstracted away in the `GetPresubmit` and `GetPostsubmit` functions ([e.g.](https://github.com/kubernetes/test-infra/blob/f31ea794ef1ee0d8d15f2b5e6e90912a25578616/prow/config/config.go#L458-L464)).
I think this duplicates logic and requires API users to provide an unnecessary field.

Reply:
If we look at [how Sub does it](https://github.com/kubernetes/test-infra/blob/c632f3a4b11b4a2adfb8fed49f079f77d2207bb6/prow/pubsub/subscriber/subscriber.go#L237-L239), there we determine whether to look at inrepoconfig or not based on whether our binary (`sub`) was started with a cache flag. I thought that it was simpler to just let clients decide where the job is defined, since they already know what the job name is.

I suppose we can just have gangway *always* look up inrepoconfig jobs on top of the static jobs --- and then from this combined list do a lookup of what job we want to run. Is this what we want here?

Reply:
Disregard my last comment. We simply check inrepoconfig if the cache pointer is valid (not nil). So yeah I think we can drop this field.

Reply:
Sub's behavior of only respecting inrepoconfig if the cache is specified is not optimal, for correctness it should also respect inrepoconfig without a cache, but that isn't implemented. Our tenants do use inrepoconfig though so we have this enabled for our Sub instance meaning we always use `GetPresubmits` and `GetPostsubmits` rather than the static config only versions.

> I suppose we can just have gangway _always_ look up inrepoconfig jobs on top of the static jobs --- and then from this combined list do a lookup of what job we want to run. Is this what we want here?

Yes, that is definitely what we want in my opinion. Otherwise we'd be expecting users to also specify where the jobs is defined which is not apparent from all places (e.g. the front end), and is not a current requirement. Also note that we will only look up inrepoconfig if it is actually enabled for the repo and in that case it is likely that the job is defined via inrepoconfig.

Reply:
Sorry, this PR is getting old and I forgot my original intent behind how things are implemented. So, the original reason why I added a JobDefinitionType_CENTRAL_REPO and JobDefinitionType_OTHER_REPO enums (and why I wrote the code so that it depended on these enums to determine whether to fetch from the static config or the inrepoconfig cache) is because I rewrote the job fetching logic to do exactly that. The Sub example is not the same because there the job fetching logic appends both the static config and the inrepoconfig cache into a single slice, like this: https://github.com/kubernetes/test-infra/blob/b79005b5d4303e0ef0f35c9bbd6b66e40a6cb5ac/prow/config/config.go#L470. Compare this to how I did it in this PR: https://github.com/kubernetes/test-infra/pull/28036/files#r1042598985.

I thought this was cleaner because there's less mutation of state (we keep static jobs and inrepoconfig jobs separate). So going back to the original discussion, if we don't have JobDefinitionType any more in our incoming message to trigger a job, we'll have to check whether the InrepoconfigSource is nil or not. This has the same effect as checking whether the JobDefinitionType is CENTRAL_REPO or OTHER_REPO.

Reply:
I don't agree that this is better. What benefit do we gain from this?  The only significant change I see is that users would now have to explicitly specify where there job configuration lives which would be worse, right?

> if we don't have JobDefinitionType any more in our incoming message to trigger a job, we'll have to check whether the InrepoconfigSource is nil or not. This has the same effect as checking whether the JobDefinitionType is CENTRAL_REPO or OTHER_REPO.

This suggests that the information is redundant and we don't need jobDefinitionType, right?

Reply:
> I don't agree that this is better. What benefit do we gain from this? The only significant change I see is that users would now have to explicitly specify where there job configuration lives which would be worse, right?

Sorry I didn't mean to suggest that we should re-include the JobDefinitionType field. I was merely explaining a slightly different behavior vs Sub in how job configurations are fetched.

> 
> > if we don't have JobDefinitionType any more in our incoming message to trigger a job, we'll have to check whether the InrepoconfigSource is nil or not. This has the same effect as checking whether the JobDefinitionType is CENTRAL_REPO or OTHER_REPO.
> 
> This suggests that the information is redundant and we don't need jobDefinitionType, right?

Correct. The only "downside" is that maybe we'll have a 3rd source of job configurations in the future beyond inrepoconfig, and at that point the boolean checks get a little messy. But this concern is premature and I don't have a strong opinion about this at this time.